### PR TITLE
REFACTORING : Streamed long polling instead of active polling (#52)

### PR DESCRIPTION
**Pull request by garnier5**
## Content

Porphyry ne recharge la page du portfolio que s'il y a eu des modifications dans Argos. 
Pour la page des items, un eventsource attend les modifications dans Argos, ce qui permet de faire une mise à jour des données (attribut et point de vue) en temps réelle.

---

## Checklist

Please check that your pull request is correct:

- Each commit:
    - [x] corresponds to a contribution that should be notified to users,
    - [x] does not generate new errors or warnings at compile or test time,
    - [x] must be attributed to its real authors (with correct GitHub IDs and [correct syntax for multiple authors](https://help.github.com/articles/creating-a-commit-with-multiple-authors/)).
- The title of a commit should:
    - [x] begin with a contribution type
        - `FEATURE` for a behaviour allowing a user to do something new,
        - `FIX` for a behaviour which has been changed in order to meet user’s expectations,
        - `TEST` when it concerns an acceptance test,
        - `PROCESS` for a change in the way the software is built, tested, deployed,
        - `DOC` when it concerns only internal documentation (however it is better to combine it with the contribution that required this documentation change),
    - [x] be followed by a colon (`:`) with one space after and no space before,
    - [x] be followed by a title (written in English) as short, as user-centered and as explicit as possible
        - If it is a feature, the title must be the user action (beginning with a verb, and please not `manage`),
        - If it is a fix, the title must describe the intended behavior (with `should`).
    - [x] ends with a reference to the corresponding ticket with the following syntax:
        - `(closes #xx)` if xx is a feature ticket (and the commit is a complete implementation),
        - `(fixes #xx)` if xx is a fix ticket (and the commit is a complete fix),
        - `(see #xx)` otherwise,
- Each committed line is:
    - [x] useful (it would not work if removed)
        - if it is a comment line, its information could not be conveyed by better variables and function naming, better code structuring, or better commit message,
    - [x] related to this very contribution (feature, fix...),
    - [x] in English (with the exception of Gherkin scenarios in French and resulting steps),
    - [x] without any typo in variable, class or function names,
    - [x] correctly indented (spaces rather than tabs, same number of characters as in the rest of the file).

### CODE REVIEW

File: src/components/itemPage/Item.jsx [hunk 1]
```diff
@@ -89,8 +89,47 @@ class Item extends Component {
     );
   }
 
+  _get_int_update_seq(update_seq){
+    return parseInt(update_seq.slice(0,update_seq.indexOf("-")));
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: Why not use the whole `update_seq`?
Please note that this string should be considered opaque (it is not supposed to be parsed).
Moreover, its inner structure could change in an upcoming version of CouchDB (it already did).

File: src/components/itemPage/Item.jsx [hunk 2]
```diff
@@ -89,8 +89,47 @@ class Item extends Component {
...
+  }
+
+  async _get_latest_update_seq() {
+    const response = await fetch('http://if05.local');
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: Please don't use constants that only works on your computer.

File: src/components/itemPage/Item.jsx [hunk 3]
```diff
@@ -89,8 +89,47 @@ class Item extends Component {
...
+    var myjson = await response.json();
+    return myjson.update_seq;
+  }
+
+  onlistening = async () => {
+    (async () => {
+      this.update_seq = await this._get_latest_update_seq();
+
+      let params = this.props.match.params;
+      this.eventSource = new EventSource(`http://if05.local:5984/argos/_changes?filter=argos/item&feed=eventsource&since=${this.update_seq}&corpus=${params.corpus}`);
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: Please don't use constants that only works on your computer.

File: src/components/itemPage/Item.jsx [hunk 4]
```diff
@@ -89,8 +89,47 @@ class Item extends Component {
...
   }
 
+  _get_int_update_seq(update_seq){
+    return parseInt(update_seq.slice(0,update_seq.indexOf("-")));
+  }
+
+  async _get_latest_update_seq() {
+    const response = await fetch('http://if05.local');
+    var myjson = await response.json();
+    return myjson.update_seq;
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: Hmm. To use the same programming style as existing code, try using `.then()` syntax more (and `await` less).

The only case when await is better is when you need to fetch several threads together.

File: src/components/itemPage/Item.jsx [hunk 5]
```diff
@@ -103,6 +142,7 @@ class Item extends Component {
       let item = data[params.corpus][params.item];
       let itemTopics = (item.topic) ? groupBy(item.topic, ['viewpoint']) : {};
       let topics=this.state.item.topic || {};
+      topics = (itemTopics.length > 0) ? topics : {};
```
Comment (benel) [inline src/components/itemPage/Item.jsx:168]: What is it for? Is it related to this REFACTORING? Or is it a FIX for a special case that was not handled?

File: src/components/portfolioPage/Portfolio.jsx [hunk 6]
```diff
@@ -55,9 +55,24 @@ class Portfolio extends Component {
     );
   }
 
+  _get_int_update_seq(update_seq){
+    return parseInt(update_seq.slice(0,update_seq.indexOf("-")));
```
Comment (benel) [inline src/components/portfolioPage/Portfolio.jsx]: Idem.

File: src/components/portfolioPage/Portfolio.jsx [hunk 7]
```diff
@@ -55,9 +55,24 @@ class Portfolio extends Component {
...
+  }
+
+  async _get_latest_update_seq() {
+    const response = await fetch('http://if05.local');
```
Comment (benel) [inline src/components/portfolioPage/Portfolio.jsx]: Idem.

File: src/components/portfolioPage/Portfolio.jsx [hunk 8]
```diff
@@ -55,9 +55,24 @@ class Portfolio extends Component {
...
+    return myjson.update_seq;
+  }
+
   componentDidMount() {
     let start=new Date().getTime();
     var self=this;
+
+    (async () => {
+      this.update_seq = await this._get_latest_update_seq()
+    })();
```
Comment (benel) [inline src/components/portfolioPage/Portfolio.jsx]: This look weird: you call an async function and do nothing with the result?

File: src/components/portfolioPage/Portfolio.jsx [hunk 9]
```diff
@@ -66,8 +81,13 @@ class Portfolio extends Component {
       let intervalTime=Math.max(10000,elapsedTime*5);
       console.log("reload every ",intervalTime);
       self._timer = setInterval(
-        () => {
-          self._fetchAll();
+        async () => {
+          var latest_seq = await this._get_latest_update_seq();
+
+          if(self._get_int_update_seq(self.update_seq)<self._get_int_update_seq(latest_seq)){
```
Comment (benel) [inline src/components/portfolioPage/Portfolio.jsx]: - Being different is enough (how could we go back to the past?).
- Be careful with eslint rules (there should be a space after `if` and spaces around binary operators such as `<` or `==`).

**Review by benel — CHANGES_REQUESTED** [review]
A nice draft, but the code needs improvements before being integrated.

File: src/components/itemPage/Item.jsx (hunk 1 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: I understand your point, but I don't see how to compare two strings without getting the number provided by the `update-seq`. 
If we compare "1432-aaa" and "931-aa" as a string, we will get a bad result.

File: src/components/itemPage/Item.jsx (hunk 3 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: Where is the constant ?

File: src/components/itemPage/Item.jsx (hunk 5 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx:168]: He's definitely part of this REFACTORING.
It fixes a noticeable problem with this update: the last deleted topic was not updated on all other client browsers.

File: src/components/itemPage/Item.jsx (hunk 1 above)
Comment (benel) [reply src/components/itemPage/Item.jsx]: We just need to check it it is different.

File: src/components/portfolioPage/Portfolio.jsx (hunk 8 above)
Comment (garnier5) [reply src/components/portfolioPage/Portfolio.jsx]: I think `componentDidMount` cannot be asynchronous.
And I needed to call this function which is asynchronous in the `componentDidMount` function.
So I used this trick.

File: src/components/itemPage/Item.jsx (hunk 3 above)
Comment (benel) [reply src/components/itemPage/Item.jsx]: `if05.local` is only defined on your computer...

File: src/components/itemPage/Item.jsx (hunk 1 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: > Being different is enough (how could we go back to the past?).

You are right.

File: src/components/portfolioPage/Portfolio.jsx (hunk 8 above)
Comment (benel) [reply src/components/portfolioPage/Portfolio.jsx]: My bad: I didn't see you modified `this.update_seq`.

However I find it risky to set it asynchronously and get it asynchronously with no synchronization between both.

File: src/components/itemPage/Item.jsx (hunk 3 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: Beginner's mistake. Am I allowed to add a constant in the configuration file (.yml) ?

Comment (benel) [reply src/components/itemPage/Item.jsx]: > Beginner's mistake. Am I allowed to add a constant in the configuration file (.yml) ?

You don't need to define a new constant. You must get `(await conf).services[0]`.

Comment (garnier5) [reply src/components/itemPage/Item.jsx]: OK, but how do we deal with users who decide to put "steatite" as their first service ?

Comment (benel) [reply src/components/itemPage/Item.jsx]: > OK, but how do we deal with users who decide to put "steatite" as their first service ?

This is not possible. The first service is the default one. It is the only one that is read-write. The other ones are read only.

Comment (benel) [reply src/components/itemPage/Item.jsx]: However, you should handle the case in which there are no `update_seq`.
Since it is not defined in the protocol, it could happen (in theory).

File: src/components/portfolioPage/Portfolio.jsx (hunk 8 above)
Comment (garnier5) [reply src/components/portfolioPage/Portfolio.jsx]: Is that better?
```
(async () => {
      let SERVICE = (await conf).services[0];
      const response = await fetch(SERVICE);
      var myjson = await response.json();
      this.update_seq = myjson.update_seq;
    })();
```

Comment (benel) [reply src/components/portfolioPage/Portfolio.jsx]: My comment was more on the global than local level: to be sure that you get the value **after** it was set (or at least assure it would not break if events do not occur in the right order).

However it is hard for me to see the whole picture of how it will look when everything is fixed. So, please, update your commit first (with all the required changes), and I will tell you if it seems OK...

**Comment by garnier5** [comment]
The changes are made, but Travis doesn't seem to be intent on testing the code...

**Comment by benel** [comment]
> The changes are made, but Travis doesn't seem to be intent on testing the code...

Vous avez l'explication dans les logs :
https://travis-ci.org/github/Hypertopic/Porphyry/builds/691004544#L243-L249

Ce sont les tests de style d'`eslint`.

**Comment by garnier5** [comment]
OK, je n'avais pas les logs lorsque j'ai écrit le message.
Cette fois, Travis ne pose plus problème.

File: src/components/itemPage/Item.jsx [hunk 10]
```diff
@@ -89,8 +90,41 @@ class Item extends Component {
...
 
+  onlistening = async () => {
+    (async () => {
+      let SERVICE = (await conf).services[0];
+      const response = await fetch(SERVICE);
+      var myjson = await response.json();
+      this.update_seq = myjson.update_seq;
+
+      let params = this.props.match.params;
+      this.eventSource = new EventSource(`${SERVICE}:5984/argos/_changes?filter=argos/item&feed=eventsource&since=${this.update_seq}&corpus=${params.corpus}`);
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: > `${SERVICE}:5984/argos`

`SERVICE` is supposed to be the Argos endpoint. 
On your computer it was `http://if05.local:5984/argos`, on the test server it is `http://argos2.test.hypertopic.org`. Relatively to this endpoint, the path is `/_changes ...`.

File: src/components/itemPage/Item.jsx [hunk 11]
```diff
@@ -89,8 +90,41 @@ class Item extends Component {
     );
   }
 
+  onlistening = async () => {
+    (async () => {
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: I guess you can remove this line now.

File: src/components/itemPage/Item.jsx [hunk 12]
```diff
@@ -89,8 +90,41 @@ class Item extends Component {
     );
   }
 
+  onlistening = async () => {
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: `onX` is usually the name of a function you give to a "handler" as a parameter.
This name is misleading since you call it directly (i.e. you trigger the action yourself, hence it should be phrased as an order).

**Review by benel — CHANGES_REQUESTED** [review]
Here are some additional advices...

File: src/components/itemPage/Item.jsx (hunk 10 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: Ah. So the code is not identical for development and production...

Comment (garnier5) [reply src/components/itemPage/Item.jsx]: Isn't it something that can be redefined in the rewrite of Argos?

Comment (garnier5) [reply src/components/itemPage/Item.jsx]: Wouldn't it be easier to create a new constant?

Comment (benel) [reply src/components/itemPage/Item.jsx]: > Isn't it something that can be redefined in the rewrite of Argos?

I don't understand: it is already.

> Ah. So the code is not identical for development and production..

It should be. It's your development constant that was erroneous.
You should have used AAAforREST (the reverse proxy with rewrite) in front of Argos.

Comment (benel) [reply src/components/itemPage/Item.jsx]: If Argos is installed as stated [here](https://github.com/Hypertopic/Argos/blob/v4/README.md), `http://localhost/` is rewritten as `http://localhost:5984/argos/_design/argos/_rewrite/`.

Therefore `http://localhost/_changes` is rewritten as `http://localhost:5984/argos/_design/argos/_rewrite/_changes` which is itself rewritten as `http://localhost:5984/argos/_changes`.

Let `SERVICE` be `http://localhost` on development environment and `http://argos2.hypertopic.org`on production, then changes API is available on `${SERVICE}/_changes` (whatever the environment).

File: src/components/itemPage/Item.jsx [hunk 13]
```diff
@@ -89,8 +90,39 @@ class Item extends Component {
...
+    this.update_seq = myjson.update_seq;
+
+    let params = this.props.match.params;
+    this.eventSource = new EventSource(`${SERVICE}/_changes?filter=argos/item&feed=eventsource&since=${this.update_seq}&corpus=${params.corpus}`);
+
+    var self=this;
+    this._fetchItem().then(() => {
+      this.eventSource.onmessage = e => {
+        var data = JSON.parse(e.data);
+        if (self.update_seq === undefined || self.update_seq !== data.seq){
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: The first clause is unnecessary (if it is undefined it will be different).

File: src/components/itemPage/Item.jsx [hunk 14]
```diff
@@ -89,8 +90,39 @@ class Item extends Component {
     );
   }
 
+  listenEventSource = async () => {
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: In this function, there are two totally different parts (on a logical and time perspectives) that appears too "tangled":

- the first call to `_fetchItems` (which personally I would have let in its original place),
- the initialization of the event source (which requires to get the `update_seq` first).

File: src/components/itemPage/Item.jsx [hunk 15]
```diff
@@ -24,7 +24,8 @@ class Item extends Component {
     super();
     this.state = {
       attributeInputValue:"",
-      item:{topic:[]}
+      item:{topic:[]},
+      update_seq:undefined
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: It is not very common to use `undefined`... Why not `null`?

File: src/components/itemPage/Item.jsx [hunk 16]
```diff
@@ -24,7 +24,8 @@ class Item extends Component {
     super();
     this.state = {
       attributeInputValue:"",
-      item:{topic:[]}
+      item:{topic:[]},
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: A small tip to avoid changing an unrelated line: adding the new element not as the last one of the list.

**Review by benel — CHANGES_REQUESTED** [review]

File: src/components/itemPage/Item.jsx (hunk 13 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: Not necessarily, data.seq can be undefined if it can't reach the service.
That's why I added it.

Comment (benel) [reply src/components/itemPage/Item.jsx]: > Not necessarily, data.seq can be undefined if it can't reach the service.
That's why I added it.

Sorry, but I really doubt that it will be sufficient to handle the case of a service that does not implement `_changes` and `update_seq`. Test it on steatite to have an idea... but I would be surprised if you didn't have an exception...

**Comment by garnier5** [comment]
Travis fait encore des siennes. Mais je ne pense pas que cela vienne de mon code...

**Comment by benel** [comment]
> Travis fait encore des siennes. Mais je ne pense pas que cela vienne de mon code...

Le site touristique de Troyes a changé de structure entre hier et ce soir... Je viens de corriger le test : cb59e64969651f10314fd2cb3d34764bb0c5125f

File: src/components/portfolioPage/Portfolio.jsx [hunk 17]
```diff
@@ -55,22 +55,47 @@ class Portfolio extends Component {
...
+              let latest_seq = data.update_seq;
+              if (self.update_seq !== latest_seq){
+                self.update_seq = latest_seq;
+                self._fetchAll();
+              }
+            });
+          },
+          intervalTime
+        );
+      });
```
Comment (benel) [inline src/components/portfolioPage/Portfolio.jsx]: Les modifications de ce fichier sont très bien et intégrables tout de suite.
Par contre, ce n'est pas du "streamed long polling" : le titre est trompeur (le titre correspond par contre à ce que vous avez fait dans l'autre fichier). 

Je serais favorable à ce que vous fassiez deux commits différents (dans deux pull requests différentes).

File: src/components/itemPage/Item.jsx [hunk 18]
```diff
@@ -89,8 +89,49 @@ class Item extends Component {
...
+      .then(data => {
+        this.update_seq = data.update_seq;
+      });
+  }
+
+  initEventSource = async () => {
+    let SERVICE = (await conf).services[0];
+    return this._fetch_update_seq().then(()=>{
+      let params = this.props.match.params;
+      this.eventSource = new EventSource(`${SERVICE}/_changes?filter=argos/item&feed=eventsource&since=${this.update_seq}&corpus=${params.corpus}`);
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: En fait, recharger les données sur un changement dans le corpus sera surtout nécessaire le jour où la réutilisation d'attributs et de valeurs sera implémentée (#156) : pour l'instant c'est juste l'item (au cas où plusieurs personnes travailleraient en même temps sur le même item). Mais OK car en plus de prévoir l'évolution future, l'écoute du corpus subsume celle de l'item.

Par contre, il **manque l'écoute des points de vue**. Cela est nécessaire si par exemple, parallèlement à l'édition d'un item, on édite la structure du point de vue dans un autre onglet.

Du coup je vous propose d'écrire dans Argos un  filtre un peu plus puissant, dans lequel on puisse décrire toute ce que l'on veut écouter. Pensez aussi à lui créer une entrée dans le fichier de **réécriture** pour que son URI soit plus simple. Ça pourrait être par exemple : 

`/events?corpus=Vitraux+-+Bénel&viewpoint= a76306e4f17ed4f79e7e481eb9a1bd06,56e092d8a6179a788c74b618b29801c0` 

voire même : 

`/events?item_corpus=Vitraux+-+Bénel&_id= a76306e4f17ed4f79e7e481eb9a1bd06,56e092d8a6179a788c74b618b29801c0"`.

Le second est plus générique et plus extensible mais peut-être plus difficile à coder. 
Une préférence ?

**Review by benel — CHANGES_REQUESTED** [review]
Très bon travail.

La partie du portfolio est intégrable tout de suite, telle quelle (mais avec un autre intitulé de commit). 

La partie de l'item doit être amendée pour gérer également la modification des points de vue qui apparaissent dans la page de l'item.

File: src/components/itemPage/Item.jsx (hunk 18 above)
Comment (garnier5) [reply src/components/itemPage/Item.jsx]: > Une préférence ?

J'ai l'impression que les deux URI sont les mêmes (à un nom d'attribut près)... Est-ce que je me trompe ?

**Comment by garnier5** [comment]
Les points de vues sont maintenant modifié dès qu'il y a un changement en base.
La partie sur le portfolio fait maintenant parti d'une nouvelle demande d'intégration (voir [ici](https://github.com/Hypertopic/Porphyry/pull/356)).

**Comment by benel** [comment]
Merci. Attention, j'ai fait du refactoring de votre code précédent. Du coup, vous avez un certain nombre de conflits de modification...

**Comment by benel** [comment]
En fait, non. C'est sur le portfolio que j'ai fait du refactoring...

File: src/components/itemPage/Item.jsx [hunk 19]
```diff
@@ -107,8 +107,55 @@ class Item extends Component {
...
+  }
+
+  initEventSource = async () => {
+    let SERVICE = (await conf).services[0];
+    return this._fetch_update_seq().then(() => {
+      let params = this.props.match.params;
+      let viewpoints = '';
+      Object.entries(this.state.item.topic).forEach(v => {
+        viewpoints += v[0] + ',';
+      });
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: Ce serait plus élégant avec`keys` plutôt que `entries` (puisque vous n'utilisez finalement que les clefs et non les valeurs) et `join` plutôt que la concaténation dans le `forEach` : 

```js
let viewpoints = Object.keys(this.state.item.topic).join(',');
```

File: src/components/itemPage/Viewpoint.jsx [hunk 20]
```diff
@@ -12,12 +12,21 @@ class Viewpoint extends React.Component {
...
     this.state = {
+      update_seq: props.update_seq,
       topics: {},
       topicInputvalue: '',
       currentSelection: '',
     };
   }
 
+  componentWillReceiveProps({update_seq}) {
+    if(update_seq){
```
Comment (benel) [inline src/components/itemPage/Viewpoint.jsx]: Dans les règles eslint du projet, après le `if ` il y a un espace (c'est ce qui le distingue d'une fonction).

File: src/components/itemPage/Item.jsx [hunk 21]
```diff
@@ -107,8 +107,55 @@ class Item extends Component {
...
+      this.eventSource = new EventSource(`${SERVICE}/events?since=${this.state.update_seq}&corpus=${params.corpus}&viewpoint=${viewpoints}`);
+    });
+  }
+
   componentDidMount() {
-    this._fetchItem();
+    if (!this.eventSource){
+      this._fetch_update_seq();
+      var self = this;
```
Comment (benel) [inline src/components/itemPage/Item.jsx]: Dans les lignes que vous ajoutez, essayez d'utiliser le plus possible `let` (qui a un comportement beaucoup plus normal que le vieux `var` de JavaScript).

**Review by benel — CHANGES_REQUESTED** [review]
Merci pour cette contribution.

Quelques améliorations de style...
